File: scripts/build_dhl_remote.py
#!/usr/bin/env python3
from __future__ import annotations

import gzip
import json
import os
import re
import tempfile
import urllib.request
from pathlib import Path

from pypdf import PdfReader

SUPPORTED = {"IT", "GR", "HR", "DE", "AT", "FR", "ES", "GB", "US", "CA", "AE", "CN", "AU"}
NUMERIC_LENGTHS = {"IT": 5, "GR": 5, "HR": 5, "DE": 5, "AT": 4, "FR": 5, "ES": 5, "US": 5, "CN": 6, "AU": 4}
CA_LETTERS = "ABCEGHJKLMNPRSTVWXYZ"
CA_INDEX = {char: index for index, char in enumerate(CA_LETTERS)}
CA_RADICES = (20, 10, 20, 10, 20, 10)
URL = "".join((
    "https://mydhl.express.dhl/content/dam/downloads/",
    "g0/remote-areas/",
    "dhl_express_remote_areas_en.pdf.coredownload.pdf",
))


def normalize(value: str) -> str:
    return re.sub(r"[^A-Z0-9]", "", value.upper())


def ca_index(value: str) -> int:
    value = normalize(value)
    if len(value) != 6:
        raise ValueError(value)
    values = (CA_INDEX[value[0]], int(value[1]), CA_INDEX[value[2]], int(value[3]), CA_INDEX[value[4]], int(value[5]))
    result = 0
    for item, radix in zip(values, CA_RADICES):
        result = result * radix + item
    return result


def merge_ranges(ranges: list[tuple[int, int]]) -> list[list[int]]:
    if not ranges:
        return []
    ranges.sort()
    merged: list[list[int]] = []
    start, end = ranges[0]
    for next_start, next_end in ranges[1:]:
        if next_start <= end + 1:
            end = max(end, next_end)
        else:
            merged.append([start, end])
            start, end = next_start, next_end
    merged.append([start, end])
    return merged


def parse_pdf(pdf: Path) -> dict:
    numeric: dict[str, list[tuple[int, int]]] = {code: [] for code in NUMERIC_LENGTHS}
    canada: list[tuple[int, int]] = []
    gb: list[list[str]] = []

    for page in PdfReader(str(pdf)).pages:
        for line in (page.extract_text() or "").splitlines():
            tokens = line.split()
            code_pos = next((i for i, token in enumerate(tokens) if token in SUPPORTED), None)
            if code_pos is None:
                continue
            code = tokens[code_pos]
            tail = tokens[code_pos + 1:]
            if code in {"CA", "GB"}:
                if len(tail) < 4:
                    continue
                start = normalize(tail[-4] + tail[-3])
                end = normalize(tail[-2] + tail[-1])
                if code == "CA":
                    try:
                        canada.append((ca_index(start), ca_index(end)))
                    except (ValueError, KeyError):
                        pass
                elif any(ch.isdigit() for ch in start) and any(ch.isdigit() for ch in end):
                    gb.append([start, end])
                continue

            expected = NUMERIC_LENGTHS.get(code)
            if expected is None or len(tail) < 2:
                continue
            start, end = normalize(tail[-2]), normalize(tail[-1])
            if (len(start) != expected or len(end) != expected) and len(tail) >= 4:
                start = normalize(tail[-4] + tail[-3])
                end = normalize(tail[-2] + tail[-1])
            if len(start) == expected and len(end) == expected and start.isdigit() and end.isdigit():
                a, b = int(start), int(end)
                numeric[code].append((min(a, b), max(a, b)))

    return {
        "version": "2026-01-04",
        "numeric": {code: {"length": NUMERIC_LENGTHS[code], "ranges": merge_ranges(ranges)} for code, ranges in numeric.items()},
        "CA": merge_ranges(canada),
        "GB": sorted({(a, b) for a, b in gb}),
    }


def obtain_pdf() -> Path:
    local = os.environ.get("DHL_REMOTE_PDF")
    if local:
        return Path(local)
    target = Path(tempfile.gettempdir()) / "dhl-remote-2026.pdf"
    request = urllib.request.Request(URL, headers={"User-Agent": "Mozilla/5.0"})
    with urllib.request.urlopen(request, timeout=180) as response:
        target.write_bytes(response.read())
    return target


def main() -> None:
    root = Path(__file__).resolve().parents[1]
    output = root / "public" / "dhl-remote-2026.json.gz"
    output.parent.mkdir(exist_ok=True)
    data = parse_pdf(obtain_pdf())
    required = set(NUMERIC_LENGTHS) | {"CA", "GB"}
    missing = [code for code in sorted(required) if not (data[code] if code in {"CA", "GB"} else data["numeric"][code]["ranges"])]
    if missing:
        raise RuntimeError("Missing DHL remote data: " + ", ".join(missing))
    payload = json.dumps(data, separators=(",", ":")).encode()
    output.write_bytes(gzip.compress(payload, compresslevel=9, mtime=0))
    print(f"Built {output} ({output.stat().st_size} bytes)")


if __name__ == "__main__":
    main()

File: scripts/build-dhl-remote.mjs
import { writeFile, readFile } from "node:fs/promises";
import { getDocument } from "pdfjs-dist/legacy/build/pdf.mjs";

const SOURCE_URL = [
  "https://mydhl.express.dhl/content/dam/downloads/",
  "g0/remote-areas/",
  "dhl_express_remote_areas_en.pdf.coredownload.pdf",
].join("");

const SUPPORTED = new Set(["IT", "GR", "HR", "DE", "AT", "FR", "ES", "GB", "US", "CA", "CN", "AU"]);
const NUMERIC_LENGTHS = { IT: 5, GR: 5, HR: 5, DE: 5, AT: 4, FR: 5, ES: 5, US: 5, CN: 6, AU: 4 };
const CA_LETTERS = "ABCEGHJKLMNPRSTVWXYZ";
const CA_RADICES = [20, 10, 20, 10, 20, 10];

const normalize = (value) => value.toUpperCase().replace(/[^A-Z0-9]/g, "");

function canadaIndex(value) {
  const zip = normalize(value);
  if (zip.length !== 6) throw new Error(`Invalid Canadian postal code: ${value}`);
  let result = 0;
  for (let index = 0; index < zip.length; index += 1) {
    const item = index % 2 === 0 ? CA_LETTERS.indexOf(zip[index]) : Number(zip[index]);
    if (item < 0 || !Number.isFinite(item)) throw new Error(`Invalid Canadian postal code: ${value}`);
    result = result * CA_RADICES[index] + item;
  }
  return result;
}

function mergeRanges(ranges) {
  if (!ranges.length) return [];
  ranges.sort((a, b) => a[0] - b[0] || a[1] - b[1]);
  const merged = [];
  let [start, end] = ranges[0];
  for (const [nextStart, nextEnd] of ranges.slice(1)) {
    if (nextStart <= end + 1) end = Math.max(end, nextEnd);
    else {
      merged.push([start, end]);
      [start, end] = [nextStart, nextEnd];
    }
  }
  merged.push([start, end]);
  return merged;
}

function pageLines(items) {
  const rows = [];
  for (const item of items) {
    if (!("str" in item) || !item.str?.trim() || !item.transform) continue;
    const x = item.transform[4];
    const y = item.transform[5];
    let row = rows.find((candidate) => Math.abs(candidate.y - y) < 1.5);
    if (!row) {
      row = { y, items: [] };
      rows.push(row);
    }
    row.items.push({ x, text: item.str.trim() });
  }
  return rows
    .sort((a, b) => b.y - a.y)
    .map((row) => row.items.sort((a, b) => a.x - b.x).map((item) => item.text).join(" "));
}

function parseLine(line, numeric, canada, gb) {
  const tokens = line.split(/\s+/).filter(Boolean);
  const codePosition = tokens.findIndex((token) => SUPPORTED.has(token));
  if (codePosition < 0) return;
  const code = tokens[codePosition];
  const tail = tokens.slice(codePosition + 1);

  if (code === "CA" || code === "GB") {
    if (tail.length < 4) return;
    const start = normalize(tail.at(-4) + tail.at(-3));
    const end = normalize(tail.at(-2) + tail.at(-1));
    if (code === "CA") {
      try {
        const a = canadaIndex(start);
        const b = canadaIndex(end);
        canada.push([Math.min(a, b), Math.max(a, b)]);
      } catch {
        // City-only rows and malformed headers are intentionally ignored.
      }
    } else if (/\d/.test(start) && /\d/.test(end)) {
      gb.push([start, end]);
    }
    return;
  }

  const expected = NUMERIC_LENGTHS[code];
  if (!expected || tail.length < 2) return;
  let start = normalize(tail.at(-2));
  let end = normalize(tail.at(-1));
  if ((start.length !== expected || end.length !== expected) && tail.length >= 4) {
    start = normalize(tail.at(-4) + tail.at(-3));
    end = normalize(tail.at(-2) + tail.at(-1));
  }
  if (start.length !== expected || end.length !== expected || !/^\d+$/.test(start + end)) return;
  const a = Number(start);
  const b = Number(end);
  numeric[code].push([Math.min(a, b), Math.max(a, b)]);
}

async function buildDatabase() {
  const response = await fetch(SOURCE_URL, { headers: { "User-Agent": "Mozilla/5.0" } });
  if (!response.ok) throw new Error(`DHL list download failed: HTTP ${response.status}`);
  const bytes = new Uint8Array(await response.arrayBuffer());
  const document = await getDocument({ data: bytes, disableWorker: true, useSystemFonts: true }).promise;

  const numeric = Object.fromEntries(Object.keys(NUMERIC_LENGTHS).map((code) => [code, []]));
  const canada = [];
  const gb = [];

  for (let pageNumber = 1; pageNumber <= document.numPages; pageNumber += 1) {
    const page = await document.getPage(pageNumber);
    const content = await page.getTextContent();
    for (const line of pageLines(content.items)) parseLine(line, numeric, canada, gb);
    if (pageNumber % 100 === 0) console.log(`Parsed ${pageNumber}/${document.numPages} pages`);
  }

  const database = {
    version: "2026-01-04",
    numeric: Object.fromEntries(Object.entries(numeric).map(([code, ranges]) => [code, {
      length: NUMERIC_LENGTHS[code],
      ranges: mergeRanges(ranges),
    }])),
    CA: mergeRanges(canada),
    GB: [...new Map(gb.map((range) => [`${range[0]}|${range[1]}`, range])).values()].sort(),
  };

  const missing = [
    ...Object.entries(database.numeric).filter(([, item]) => !item.ranges.length).map(([code]) => code),
    ...(database.CA.length ? [] : ["CA"]),
    ...(database.GB.length ? [] : ["GB"]),
  ];
  if (missing.length) throw new Error(`DHL remote data missing for: ${missing.join(", ")}`);
  return database;
}

async function patchSource() {
  const mainPath = new URL("../src/main.tsx", import.meta.url);
  let mainSource = await readFile(mainPath, "utf8");
  if (!mainSource.includes('import "./remoteRuntime";')) {
    mainSource = mainSource.replace('import "./dhlRuntime";', 'import "./dhlRuntime";\nimport "./remoteRuntime";');
    await writeFile(mainPath, mainSource);
  }
}

const database = await buildDatabase();
await writeFile(new URL("../public/dhl-remote-2026.json", import.meta.url), JSON.stringify(database));
await patchSource();
console.log("DHL 2026 remote-area database created.");
